File: backend/app/models/backtesting.py
"""Backtesting ORM models — BacktestRun, BacktestTrade, BacktestMetrics."""

from datetime import datetime

from sqlalchemy import String, Float, Integer, DateTime, JSON, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class BacktestRun(Base):
    """A single backtesting run — configuration and results."""

    __tablename__ = "backtest_runs"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    instrument: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    timeframe: Mapped[str] = mapped_column(String(10), nullable=False)
    start_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    end_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending", index=True)
    total_bars: Mapped[int] = mapped_column(Integer, default=0)
    config_json: Mapped[str | None] = mapped_column(String(5000), nullable=True)
    metrics_json: Mapped[str | None] = mapped_column(String(5000), nullable=True)
    equity_curve_json: Mapped[str | None] = mapped_column(String(20000), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow,
    )


class BacktestTrade(Base):
    """A completed trade recorded during backtesting."""

    __tablename__ = "backtest_trades"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    run_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("backtest_runs.id"), nullable=False, index=True,
    )
    entry_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    exit_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    direction: Mapped[str] = mapped_column(String(10), nullable=False)
    quantity: Mapped[int] = mapped_column(Integer, default=1)
    entry_price: Mapped[float] = mapped_column(Float, nullable=False)
    exit_price: Mapped[float] = mapped_column(Float, nullable=False)
    stop_price: Mapped[float] = mapped_column(Float, default=0.0)
    risk: Mapped[float] = mapped_column(Float, default=0.0)
    r_multiple: Mapped[float] = mapped_column(Float, default=0.0)
    pnl: Mapped[float] = mapped_column(Float, default=0.0)
    duration_seconds: Mapped[int] = mapped_column(Integer, default=0)
    exit_reason: Mapped[str] = mapped_column(String(50), default="")
    strategy_version: Mapped[str] = mapped_column(String(20), default="1.0.0")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow,
    )


class BacktestMetrics(Base):
    """Aggregated performance metrics for a backtest run."""

    __tablename__ = "backtest_metrics"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    run_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("backtest_runs.id"), nullable=False, unique=True,
    )

    # P&L
    net_profit: Mapped[float] = mapped_column(Float, default=0.0)
    gross_profit: Mapped[float] = mapped_column(Float, default=0.0)
    gross_loss: Mapped[float] = mapped_column(Float, default=0.0)

    # Trade counts
    total_trades: Mapped[int] = mapped_column(Integer, default=0)
    winning_trades: Mapped[int] = mapped_column(Integer, default=0)
    losing_trades: Mapped[int] = mapped_column(Integer, default=0)
    breakeven_trades: Mapped[int] = mapped_column(Integer, default=0)

    # Rates
    win_rate: Mapped[float] = mapped_column(Float, default=0.0)
    loss_rate: Mapped[float] = mapped_column(Float, default=0.0)

    # Ratios
    profit_factor: Mapped[float] = mapped_column(Float, default=0.0)

    # Averages
    average_win: Mapped[float] = mapped_column(Float, default=0.0)
    average_loss: Mapped[float] = mapped_column(Float, default=0.0)
    average_r: Mapped[float] = mapped_column(Float, default=0.0)

    # Expectancy
    expectancy: Mapped[float] = mapped_column(Float, default=0.0)

    # Drawdown
    max_drawdown: Mapped[float] = mapped_column(Float, default=0.0)
    max_drawdown_pct: Mapped[float] = mapped_column(Float, default=0.0)

    # Streaks
    max_consecutive_wins: Mapped[int] = mapped_column(Integer, default=0)
    max_consecutive_losses: Mapped[int] = mapped_column(Integer, default=0)

    # Duration
    average_trade_duration_seconds: Mapped[float] = mapped_column(Float, default=0.0)

    # Directional
    long_trades: Mapped[int] = mapped_column(Integer, default=0)
    long_wins: Mapped[int] = mapped_column(Integer, default=0)
    long_pnl: Mapped[float] = mapped_column(Float, default=0.0)
    short_trades: Mapped[int] = mapped_column(Integer, default=0)
    short_wins: Mapped[int] = mapped_column(Integer, default=0)
    short_pnl: Mapped[float] = mapped_column(Float, default=0.0)

    # Extremes
    largest_winner: Mapped[float] = mapped_column(Float, default=0.0)
    largest_loser: Mapped[float] = mapped_column(Float, default=0.0)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow,
    )

File: frontend/src/api/backtesting.ts
/** Backtesting API client. */

const BASE = "/api/v1";

/* ─── Types ──────────────────────────────────────────── */

export interface BacktestMetrics {
  net_profit: number;
  gross_profit: number;
  gross_loss: number;
  total_trades: number;
  winning_trades: number;
  losing_trades: number;
  breakeven_trades: number;
  win_rate: number;
  loss_rate: number;
  profit_factor: number;
  average_win: number;
  average_loss: number;
  average_r: number;
  expectancy: number;
  max_drawdown: number;
  max_drawdown_pct: number;
  max_consecutive_wins: number;
  max_consecutive_losses: number;
  average_trade_duration_seconds: number;
  long_trades: number;
  long_wins: number;
  long_pnl: number;
  short_trades: number;
  short_wins: number;
  short_pnl: number;
  largest_winner: number;
  largest_loser: number;
  // Sprint 3
  sharpe_ratio: number;
  sortino_ratio: number;
  annual_return_pct: number;
  monthly_returns: MonthlyReturn[];
  max_drawdown_duration_days: number;
  recovery_factor: number;
}

export interface MonthlyReturn {
  month: string;
  return_pct: number;
  trades: number;
  pnl: number;
}

export interface EquityPoint {
  trade_index: number;
  timestamp: string | null;
  account_balance: number;
  equity: number;
  drawdown: number;
  drawdown_pct: number;
  peak_equity: number;
}

export interface BacktestTrade {
  trade_id: string;
  entry_time: string | null;
  exit_time: string | null;
  direction: string;
  quantity: number;
  entry_price: number;
  exit_price: number;
  stop_price: number;
  risk: number;
  r_multiple: number;
  pnl: number;
  duration_seconds: number;
  exit_reason: string;
  strategy_version: string;
}

export interface BacktestConfig {
  instrument: string;
  timeframe: string;
  start_time: string | null;
  end_time: string | null;
  replay_mode: string;
  strategy: string;
  strategy_params: Record<string, unknown>;
  initial_balance: number;
  commission_per_contract: number;
  slippage_ticks: number;
}

export interface BacktestResult {
  run_id: string;
  config: BacktestConfig;
  trades: BacktestTrade[];
  equity_curve: EquityPoint[];
  metrics: BacktestMetrics;
  errors: string[];
}

export interface RunListItem {
  id: number;
  instrument: string;
  timeframe: string;
  start_time: string;
  end_time: string;
  status: string;
  created_at: string;
}

/* ─── Helpers ────────────────────────────────────────── */

async function fetchJson<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, init);
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`HTTP ${res.status}: ${text || res.statusText}`);
  }
  return res.json();
}

/* ─── Run Backtest ───────────────────────────────────── */

export async function runBacktest(params: {
  instrument: string;
  timeframe: string;
  start_time: string;
  end_time: string;
  strategy: string;
  bars_json: string;
  initial_balance?: number;
  commission?: number;
  slippage?: number;
}): Promise<{
  run_id: number;
  status: string;
  metrics: BacktestMetrics;
  trade_count: number;
  equity_curve: EquityPoint[];
}> {
  const q = new URLSearchParams({
    instrument: params.instrument,
    timeframe: params.timeframe,
    start_time: params.start_time,
    end_time: params.end_time,
    strategy: params.strategy,
    bars_json: params.bars_json,
    initial_balance: String(params.initial_balance ?? 100000),
    commission: String(params.commission ?? 2.5),
    slippage: String(params.slippage ?? 1),
  });
  return fetchJson(`${BASE}/backtesting/run?${q}`, { method: "POST" });
}

/* ─── Run Dry Backtest ───────────────────────────────── */

export async function runDryBacktest(params: {
  instrument: string;
  timeframe: string;
  start_time: string;
  end_time: string;
  strategy: string;
  bars_json: string;
  initial_balance?: number;
  commission?: number;
  slippage?: number;
}): Promise<BacktestResult> {
  const q = new URLSearchParams({
    instrument: params.instrument,
    timeframe: params.timeframe,
    start_time: params.start_time,
    end_time: params.end_time,
    strategy: params.strategy,
    bars_json: params.bars_json,
    initial_balance: String(params.initial_balance ?? 100000),
    commission: String(params.commission ?? 2.5),
    slippage: String(params.slippage ?? 1),
  });
  return fetchJson(`${BASE}/backtesting/run-dry?${q}`, { method: "POST" });
}

/* ─── List Runs ──────────────────────────────────────── */

export async function listRuns(instrument?: string): Promise<{
  count: number;
  runs: RunListItem[];
}> {
  const params = instrument ? `?instrument=${instrument}` : "";
  return fetchJson(`${BASE}/backtesting/runs${params}`);
}

/* ─── Get Run ────────────────────────────────────────── */

export async function getRun(runId: number): Promise<{
  id: number;
  instrument: string;
  timeframe: string;
  start_time: string;
  end_time: string;
  status: string;
  total_bars: number;
  metrics_json: string;
  equity_curve_json: string;
  created_at: string;
}> {
  return fetchJson(`${BASE}/backtesting/runs/${runId}`);
}
